apps/src/code-studio/components/ExportAllowedDialog.jsx

```diff
@@ -0,0 +1,744 @@
+      '_blank'
+    );
+  };
+
+  async publishExpoExport() {
+    const {expoUri} = this.state;
+    if (expoUri) {
+      // We have already have exported to Expo
+      return;
+    }
+    const {exportApp, projectUpdatedAt} = this.props;
+    this.setState({
+      exporting: true,
+      exportProjectUpdatedAt: projectUpdatedAt
+    });
+    try {
+      const {expoUri, expoSnackId, iconUri, splashImageUri} = await exportApp({
+        mode: 'expoPublish'
+      });
+      this.setState({
+        exporting: false,
+        exportError: null,
+        expoUri,
+        expoSnackId,
+        iconUri,
+        splashImageUri
+      });
+    } catch (e) {
+      this.setState({
+        exporting: false,
+        exportProjectUpdatedAt: null,
+        expoUri: null,
+        expoSnackId: null,
+        exportError:
+          'Failed to publish project to Expo. Please try again later.'
+      });
+    }
+  }
+
+  async publishAndGenerateApk() {
+    if (this.state.apkUri) {
+      // We have already have generated an APK
+      return;
+    }
+
+    await this.publishExpoExport();
+
+    const {exportApp} = this.props;
+    const {expoSnackId, iconUri, splashImageUri} = this.state;
+
+    if (!expoSnackId) {
+      // We failed to generate a snackId
+      return;
+    }
+
+    this.setState({generatingApk: true});
+    try {
+      const apkUri = await exportApp({
+        mode: 'expoGenerateApk',
+        expoSnackId,
+        iconUri,
+        splashImageUri
+      });
+      this.setState({
+        generatingApk: false,
+        apkError: null,
+        apkUri
+      });
+    } catch (e) {
+      this.setState({
+        generatingApk: false,
+        apkError: 'Failed to create Android app. Please try again later.'
+      });
+    }
+  }
+
+  onActionButton = () => {
+    const {screen} = this.state;
+
+    switch (screen) {
+      case 'intro':
+        return this.setState({screen: 'platform'});
+      // case 'export':
+      //   return this.setState({screen: 'platform'});
+      case 'platform':
+        return this.setState({screen: 'publish'});
+      // return this.setState({screen: 'icon'});
+      // case 'icon':
+      //   return this.setState({screen: 'publish'});
+      case 'publish':
+        this.publishAndGenerateApk();
+        return this.setState({screen: 'generating'});
+      case 'generating':
+        return this.close();
+      default:
+        throw new Error(`ExportAllowedDialog: Unexpected screen: ${screen}`);
+    }
+  };
+
+  renderMainContent() {
+    const {screen} = this.state;
+
+    switch (screen) {
+      case 'intro':
+        return this.renderIntroPage();
+      // case 'export':
+      //   return this.renderExportPage();
+      case 'platform':
+        return this.renderPlatformPage();
+      // case 'icon':
+      //   return this.renderIconPage();
+      case 'publish':
+        return this.renderPublishPage();
+      case 'generating':
+        return this.renderGeneratingPage();
```

Comment:
Calling out a personal preference for subcomponents over helper render methods, but that's easy to fix later.

Reply:
Understood, I'd be happy to revisit this later.